README.rst
dateglob
========

``dateglob`` converts a set of dates into a list of globs. For example:

>>> import dateglob; from datetime import date, timedelta
>>> # build list of dates from 2009-12-31 thru 2011-02-01
>>> dates = [date(2009, 12, 31) + timedelta(i) for i in xrange(1+365+31+1)]
>>> dateglob.strftime(dates, '%Y-%m-%d')
['2009-12-31', '2010-*-*', '2011-01-*', '2011-02-01']

The original use case for this library was to generate compact command lines
for command that take daily log files as input, for example:

>>> args += dateglob.strftime(dates, '/logs/foo/%Y/%m/%d/*.gz')

``dateglob.strftime()`` handles all of the standard arguments to ``datetime.strftime()``. Currently, it only does something special with
full months, years, and full ten-day periods (we don't glob weeks).

* source: <http://github.com/Yelp/dateglob>
* documentation: <http://packages.python.org/dateglob/>
* datetime.strftime(): <http://docs.python.org/library/datetime.html#strftime-and-strptime-behavior>


CHANGES.txt
v0.3, 2016-07-11 -- Python 3.5 compatibility!

v0.2, 2016-05-1* -- Ten-day period globs!

v0.1, 2010-06-* -- Initial release!


tests/dateglob_test.py
from datetime import date, timedelta

from testify import TestCase
from testify import assert_equal
from testify import assert_raises
from testify import run

from dateglob import strftime


def m(year, month):
    """Return a list of days in the given month"""
    d = start = date(year, month, 1)
    while d.month == start.month:
        yield d
        d += timedelta(1)


def y(year):
    """Return a list of days in the given year"""
    d = start = date(year, 1, 1)
    while d.year == start.year:
        yield d
        d += timedelta(1)


def t(year, month, day, period=11):
    """Return a list of days in the given year"""
    d = date(year, month, day)
    for i in range(period):
        yield d
        d += timedelta(1)


class TestStrftime(TestCase):

    def test_empty(self):
        assert_equal(strftime([], '%Y-%m-%d'), [])
        assert_equal(strftime((), '%Y-%m-%d'), [])

    def test_empty_with_bad_format(self):
        # don't look at format string if no dates
        assert_equal(strftime([], object()), [])

    def test_non_strftime_format(self):
        assert_equal(strftime(y(2010), 'foo'), ['foo'])

        # don't look at dates at all if it's not a format string
        assert_equal(strftime([object()], 'foo'), ['foo'])

    def test_month_globbing(self):
        assert_equal(strftime(m(2010, 6), '%Y-%m-%d'), ['2010-06-*'])

        # %b, and %B are okay too, but are locale-specific
        assert_equal(strftime(m(2010, 6), '%b %d, %Y'),
                     [date(2010, 6, 1).strftime('%b *, 2010')])
        assert_equal(strftime(m(2010, 6), '%B %d, %Y'),
                     [date(2010, 6, 1).strftime('%B *, 2010')])

    def test_month_no_globbing(self):
    	# can't glob day-of-year (%j)
        assert_equal(strftime(m(2010, 6), '%Y-%j'),
                     [d.strftime('%Y-%j') for d in m(2010, 6)])

        # can't glob day-of-week (%a, %A, %W)
        assert_equal(strftime(m(2010, 6), '%Y-%a'),
                     sorted(set(d.strftime('%Y-%a') for d in m(2010, 6))))
        assert_equal(strftime(m(2010, 6), '%Y-%A'),
                     sorted(set(d.strftime('%Y-%A') for d in m(2010, 6))))
        assert_equal(strftime(m(2010, 6), '%Y-%w'),
                     sorted(set(d.strftime('%Y-%w') for d in m(2010, 6))))

        # can't glob whole date (%c, %x)
        assert_equal(strftime(m(2010, 6), '%c'),
                     sorted(d.strftime('%c') for d in m(2010, 6)))
        assert_equal(strftime(m(2010, 6), '%x'),
                     sorted(d.strftime('%x') for d in m(2010, 6)))

    def test_year_globbing(self):
        assert_equal(strftime(y(2010), '%Y-%m-%d'), ['2010-*-*'])
        assert_equal(strftime(y(2010), '%b %d, %Y'), ['* *, 2010'])
        assert_equal(strftime(y(2010), '%B %d, %Y'), ['* *, 2010'])

        # old two-digit year (%y)
        assert_equal(strftime(y(2010), '%y-%m-%d'), ['10-*-*'])

        # can handle day-of-week and day-of-year
        assert_equal(strftime(y(2010), '%Y: %a, %A, %j, %w'),
                     ['2010: *, *, *, *'])

    def test_year_no_globbing(self):
        # can't glob whole date (%c, %x)
        assert_equal(strftime(y(2010), '%c'),
                     sorted(d.strftime('%c') for d in y(2010)))
        assert_equal(strftime(y(2010), '%x'),
                     sorted(d.strftime('%x') for d in y(2010)))

    def test_ten_globbing(self):
        # Test first ten
        assert_equal(strftime(t(2016, 5, 1), '%Y-%m-%d'),
                     ['2016-05-0*', '2016-05-10', '2016-05-11'])
        # Test second ten
        assert_equal(strftime(t(2016, 5, 10), '%Y-%m-%d'),
                     ['2016-05-1*', '2016-05-20'])
        # Test third ten
        assert_equal(strftime(t(2016, 5, 20), '%Y-%m-%d'),
                     ['2016-05-2*', '2016-05-30'])

        # Test fourth ten on 30-day month
        assert_equal(strftime(t(2016, 4, 30, period=2), '%Y-%m-%d'),
                     ['2016-04-3*', '2016-05-01'])
        # Test fourth ten on 31-day month with all days
        assert_equal(strftime(t(2016, 1, 30, period=2), '%Y-%m-%d'),
                     ['2016-01-3*'])
        # Test fourth ten on 31-day month with one of the 2 days
        assert_equal(strftime(t(2016, 1, 30, period=1), '%Y-%m-%d'),
                     ['2016-01-30'])
        # Test fourth ten on 31-day month with one of the 2 days
        assert_equal(strftime(t(2016, 1, 31, period=1), '%Y-%m-%d'),
                     ['2016-01-31'])

        # Test February of Leap Year
        assert_equal(strftime(t(2016, 2, 20, period=10), '%Y-%m-%d'),
                     ['2016-02-2*'])
        # Test February of Normal Year
        assert_equal(strftime(t(2015, 2, 20, period=9), '%Y-%m-%d'),
                     ['2015-02-2*'])

    def test_readme(self):
        # good test of a range of dates
        dates = [date(2009, 12, 31) + timedelta(i) for i in xrange(1+365+31+1)]
        assert_equal(strftime(dates, '%Y-%m-%d'),
                     ['2009-12-31', '2010-*-*', '2011-01-*', '2011-02-01'])

    def test_sorting(self):
        # another test, of a discontinuous range of dates
        dates = ([date(2010, 6, 6)] +
                 list(m(2007, 7)) +
                 list(y(2011)) +
                 [date(2007, 5, 6)])

        assert_equal(strftime(dates, '%Y-%m-%d'),
                     ['2007-05-06', '2007-07-*', '2010-06-06', '2011-*-*'])

        # sorting is just alphabetical, not by date
        assert_equal(strftime(dates, '%m/%d/%y'),
                     ['*/*/11', '05/06/07', '06/06/10', '07/*/07'])

    def test_no_duplicates(self):
        dates = [date(2010, 6, 6)] * 10000
        assert_equal(strftime(dates, '%Y-%m-%d'), ['2010-06-06'])

    def test_time_fields_are_globbed(self):
        assert_equal(
            strftime([date(2010, 6, 6)], '%Y-%m-%dT%H:%M:%SZ'),
            ['2010-06-06T*:*:*Z'])

        assert_equal(
            strftime(y(2010), '%Y-%m-%d %f%I%p%S%X%z%Z'),
            ['2010-*-* *'])

    def test_no_double_wildcard(self):
        assert_equal(strftime(y(2010), '%Y%m%d'), ['2010*'])

    def test_percent_escaping(self):
        assert_equal(strftime(y(2011), '110%%'), ['110%'])

        # don't incorrectly grab % out of %% to do globbing
        assert_equal(strftime(y(2011), '%m %%m %%%m'), ['* %m %*'])

        # catch invalid strftime string
        assert_raises(ValueError, strftime, y(2011), '110%')


if __name__ == '__main__':
    run()


tests/test_dateglob.py
from datetime import date
from datetime import timedelta
from unittest import TestCase

from dateglob import strftime


def m(year, month):
    """Return a list of days in the given month"""
    d = start = date(year, month, 1)
    while d.month == start.month:
        yield d
        d += timedelta(1)


def y(year):
    """Return a list of days in the given year"""
    d = start = date(year, 1, 1)
    while d.year == start.year:
        yield d
        d += timedelta(1)


def t(year, month, day, period=11):
    """Return a list of days in the given year"""
    d = date(year, month, day)
    for i in range(period):
        yield d
        d += timedelta(1)


class TestStrftime(TestCase):

    def test_empty(self):
        self.assertEqual(strftime([], '%Y-%m-%d'), [])
        self.assertEqual(strftime((), '%Y-%m-%d'), [])

    def test_empty_with_bad_format(self):
        # don't look at format string if no dates
        self.assertEqual(strftime([], object()), [])

    def test_non_strftime_format(self):
        self.assertEqual(strftime(y(2010), 'foo'), ['foo'])

        # don't look at dates at all if it's not a format string
        self.assertEqual(strftime([object()], 'foo'), ['foo'])

    def test_month_globbing(self):
        self.assertEqual(strftime(m(2010, 6), '%Y-%m-%d'), ['2010-06-*'])

        # %b, and %B are okay too, but are locale-specific
        self.assertEqual(strftime(m(2010, 6), '%b %d, %Y'),
                     [date(2010, 6, 1).strftime('%b *, 2010')])
        self.assertEqual(strftime(m(2010, 6), '%B %d, %Y'),
                     [date(2010, 6, 1).strftime('%B *, 2010')])

    def test_month_no_globbing(self):
        # can't glob day-of-year (%j)
        self.assertEqual(strftime(m(2010, 6), '%Y-%j'),
                     [d.strftime('%Y-%j') for d in m(2010, 6)])

        # can't glob day-of-week (%a, %A, %W)
        self.assertEqual(strftime(m(2010, 6), '%Y-%a'),
                     sorted(set(d.strftime('%Y-%a') for d in m(2010, 6))))
        self.assertEqual(strftime(m(2010, 6), '%Y-%A'),
                     sorted(set(d.strftime('%Y-%A') for d in m(2010, 6))))
        self.assertEqual(strftime(m(2010, 6), '%Y-%w'),
                     sorted(set(d.strftime('%Y-%w') for d in m(2010, 6))))

        # can't glob whole date (%c, %x)
        self.assertEqual(strftime(m(2010, 6), '%c'),
                     sorted(d.strftime('%c') for d in m(2010, 6)))
        self.assertEqual(strftime(m(2010, 6), '%x'),
                     sorted(d.strftime('%x') for d in m(2010, 6)))

    def test_year_globbing(self):
        self.assertEqual(strftime(y(2010), '%Y-%m-%d'), ['2010-*-*'])
        self.assertEqual(strftime(y(2010), '%b %d, %Y'), ['* *, 2010'])
        self.assertEqual(strftime(y(2010), '%B %d, %Y'), ['* *, 2010'])

        # old two-digit year (%y)
        self.assertEqual(strftime(y(2010), '%y-%m-%d'), ['10-*-*'])

        # can handle day-of-week and day-of-year
        self.assertEqual(strftime(y(2010), '%Y: %a, %A, %j, %w'),
                     ['2010: *, *, *, *'])

    def test_year_no_globbing(self):
        # can't glob whole date (%c, %x)
        self.assertEqual(strftime(y(2010), '%c'),
                     sorted(d.strftime('%c') for d in y(2010)))
        self.assertEqual(strftime(y(2010), '%x'),
                     sorted(d.strftime('%x') for d in y(2010)))

    def test_ten_globbing(self):
        # Test first ten
        self.assertEqual(strftime(t(2016, 5, 1), '%Y-%m-%d'),
                     ['2016-05-0*', '2016-05-10', '2016-05-11'])
        # Test second ten
        self.assertEqual(strftime(t(2016, 5, 10), '%Y-%m-%d'),
                     ['2016-05-1*', '2016-05-20'])
        # Test third ten
        self.assertEqual(strftime(t(2016, 5, 20), '%Y-%m-%d'),
                     ['2016-05-2*', '2016-05-30'])

        # Test fourth ten on 30-day month
        self.assertEqual(strftime(t(2016, 4, 30, period=2), '%Y-%m-%d'),
                     ['2016-04-3*', '2016-05-01'])
        # Test fourth ten on 31-day month with all days
        self.assertEqual(strftime(t(2016, 1, 30, period=2), '%Y-%m-%d'),
                     ['2016-01-3*'])
        # Test fourth ten on 31-day month with one of the 2 days
        self.assertEqual(strftime(t(2016, 1, 30, period=1), '%Y-%m-%d'),
                     ['2016-01-30'])
        # Test fourth ten on 31-day month with one of the 2 days
        self.assertEqual(strftime(t(2016, 1, 31, period=1), '%Y-%m-%d'),
                     ['2016-01-31'])

        # Test February of Leap Year
        self.assertEqual(strftime(t(2016, 2, 20, period=10), '%Y-%m-%d'),
                     ['2016-02-2*'])
        # Test February of Normal Year
        self.assertEqual(strftime(t(2015, 2, 20, period=9), '%Y-%m-%d'),
                     ['2015-02-2*'])

    def test_readme(self):
        # good test of a range of dates
        dates = [date(2009, 12, 31) + timedelta(i) for i in range(1+365+31+1)]
        self.assertEqual(strftime(dates, '%Y-%m-%d'),
                     ['2009-12-31', '2010-*-*', '2011-01-*', '2011-02-01'])

    def test_sorting(self):
        # another test, of a discontinuous range of dates
        dates = ([date(2010, 6, 6)] +
                 list(m(2007, 7)) +
                 list(y(2011)) +
                 [date(2007, 5, 6)])

        self.assertEqual(strftime(dates, '%Y-%m-%d'),
                     ['2007-05-06', '2007-07-*', '2010-06-06', '2011-*-*'])

        # sorting is just alphabetical, not by date
        self.assertEqual(strftime(dates, '%m/%d/%y'),
                     ['*/*/11', '05/06/07', '06/06/10', '07/*/07'])

    def test_no_duplicates(self):
        dates = [date(2010, 6, 6)] * 10000
        self.assertEqual(strftime(dates, '%Y-%m-%d'), ['2010-06-06'])

    def test_time_fields_are_globbed(self):
        self.assertEqual(
            strftime([date(2010, 6, 6)], '%Y-%m-%dT%H:%M:%SZ'),
            ['2010-06-06T*:*:*Z'])

        self.assertEqual(
            strftime(y(2010), '%Y-%m-%d %f%I%p%S%X%z%Z'),
            ['2010-*-* *'])

    def test_no_double_wildcard(self):
        self.assertEqual(strftime(y(2010), '%Y%m%d'), ['2010*'])

    def test_percent_escaping(self):
        self.assertEqual(strftime(y(2011), '110%%'), ['110%'])

        # don't incorrectly grab % out of %% to do globbing
        self.assertEqual(strftime(y(2011), '%m %%m %%%m'), ['* %m %*'])

        # catch invalid strftime string
        self.assertRaises(ValueError, strftime, y(2011), '110%')


dateglob.py
# Copyright 2011 Yelp
# Copyright 2016 Patrick Boocock, Yelp
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Convert a set of dates into a compact list of globs. For example:

>>> import dateglob; from datetime import date, timedelta
>>> # build list of dates from 2009-12-31 thru 2011-02-01
>>> dates = [date(2009, 12, 31) + timedelta(i) for i in xrange(1+365+31+1)]
>>> dateglob.strftime(dates, '%Y-%m-%d')
['2009-12-31', '2010-*-*', '2011-01-*', '2011-02-01']

The original use case for this library was to generate compact command lines
for command that take daily log files as input, for example:

>>> args += dateglob.strftime(dates, '/logs/foo/%Y/%m/%d/*.gz')
"""
import calendar
from collections import defaultdict
import datetime
import re
import six


__author__ = 'David Marin <dave@yelp.com>'

__credits__ = [
    'Patrick Boocock <pboocock@jawbone.com>'
]

__version__ = '0.3'

__all__ = ['strftime']


# match directives in a strftime format string (e.g. '%Y-%m-%d')
STRFTIME_FIELD_RE = re.compile('%(.)')

# used to compact multiple adjacent wildcards into a single wildcard.
MULTIPLE_WILDCARD_RE = re.compile(r'\*+')

# all fields handled by the default version of strftime; see
# http://docs.python.org/library/datetime.html#strftime-and-strptime-behavior

# divisions smaller than a day and time zones; always converted to *
TIME_FIELDS = 'fHIMpSXzZ'

DAY_OF_WEEK_FIELDS = 'aAw'

DAY_OF_MONTH_FIELDS = 'd'

DAY_OF_YEAR_FIELDS = 'j'

# week number in year, with weeks starting on Sunday
SUNDAY_WEEK_OF_YEAR_FIELDS = 'U'

# week number in year, with weeks starting on Monday
MONDAY_WEEK_OF_YEAR_FIELDS = 'W'

# month names and numbers
MONTH_FIELDS = 'bBm'

YEAR_FIELDS = 'yY'

# fields that describe the whole date at once (make globbing impossible)
WHOLE_DATE_FIELDS = 'cx'

# The "%%" format; always leave these alone
LITERAL_FIELDS = '%'

# fields that can be replaced with * if we have all days in a year
YEAR_GLOB_FIELDS = (TIME_FIELDS +
                    DAY_OF_WEEK_FIELDS +
                    DAY_OF_MONTH_FIELDS +
                    DAY_OF_YEAR_FIELDS +
                    SUNDAY_WEEK_OF_YEAR_FIELDS +
                    MONDAY_WEEK_OF_YEAR_FIELDS +
                    MONTH_FIELDS)

# fields that preclude globbing even if we have all days in a year
YEAR_BAD_FIELDS = WHOLE_DATE_FIELDS

# fields that can be replaced with * if we have all days in a month
MONTH_GLOB_FIELDS = (TIME_FIELDS +
                     DAY_OF_MONTH_FIELDS)

# fields that preclude globbing even if we have all days in a month
MONTH_BAD_FIELDS = (WHOLE_DATE_FIELDS +
                     DAY_OF_YEAR_FIELDS +
                     DAY_OF_WEEK_FIELDS +
                     SUNDAY_WEEK_OF_YEAR_FIELDS +
                     MONDAY_WEEK_OF_YEAR_FIELDS)

TEN_LENGTH = {
    0: 9,
    1: 10,
    2: 10
}


def strftime(dates, format):
    """Format a sequence of dates, using ``*`` when possible.

    For example, if *dates* contains all days in June 2011, and
    *format* is ``'%Y/%m/%d'``, we return ``['2011/06/*']``

    Divisions smaller than a day (e.g. ``%H``) and time zone fields
    (e.g. ``%Z``) are always converted to ``*``.

    Currently, we only do something special with full months, years, and full
    ten-day periods (we don't glob weeks).

    :param dates: a sequence of :py:class:`datetime.date` objects
    :param format: a :py:func:`~datetime.date.strftime` format string (see http://docs.python.org/library/datetime.html#strftime-and-strptime-behavior)
    :return: a list of strings corresponding to strftime-formatted dates, using ``*`` wherever possible. These will be distinct (no duplicates) and in alphabetical order.
    """
    # handle special cases quickly
    if not dates:
        return []
    elif not '%' in format:
        # don't use STRFTIME_FIELD_RE to check because we want to catch
        # malformed format strings as well.
        return [format]

    results = set()

    # year globbing
    if not has_fields(format, YEAR_BAD_FIELDS):
        full_years, dates = extract_full_years(dates)
        if full_years:
            year_glob = glob_fields(format, YEAR_GLOB_FIELDS)
            for year in full_years:
                results.add(datetime.date(year, 1, 1).strftime(year_glob))

    # month globbing
    if not has_fields(format, MONTH_BAD_FIELDS):
        full_months, dates = extract_full_months(dates)
        if full_months:
            month_glob = glob_fields(format, MONTH_GLOB_FIELDS)
            for year, month in full_months:
                results.add(datetime.date(year, month, 1).strftime(month_glob))

    # tens globbing
    if not has_fields(format, MONTH_BAD_FIELDS):
        full_tens, dates = extract_full_tens(dates)
        if full_tens:
            for year, month, ten in full_tens:
                tens_glob = glob_fields(format, MONTH_GLOB_FIELDS, day_str=str(ten))
                results.add(datetime.date(year, month, 1).strftime(tens_glob))

    # everything else
    for day in dates:
        day_glob = glob_fields(format, TIME_FIELDS)
        results.add(day.strftime(day_glob))

    return sorted(results)


def has_fields(format, fields):
    """Check a format string for fields of a given type.
    :param format: a :py:func:`~datetime.date.strftime` format string
    :type fields: str
    :param fields: one-character field types to look for
    :rtype: bool
    """
    return bool(set(STRFTIME_FIELD_RE.findall(format)) & set(fields))


def glob_fields(format, fields, day_str=''):
    """Replace fields in a format string with ``*``. Adjacent stars (`**`)
    will be merged into a single star.

    :param format: a :py:func:`~datetime.date.strftime` format string
    :type fields: str
    :param fields: one-character field types to replace with ``*``
    :type day_str: str
    :param day_str: can be substituted with ten-day period prefix as str
    :rtype: bool
    """
    format = STRFTIME_FIELD_RE.sub(
        lambda m: '%s*' % day_str if m.group(1) in fields else m.group(0),
        format)
    format = MULTIPLE_WILDCARD_RE.sub('*', format)
    return format


def extract_full_years(dates):
    """Find if there are any years where every day of the year is in *dates*.

    Returns ``full_years, other_dates``: *years* is a list of year
    numbers that *dates* contains all the days of, and *other_dates* is
    everything else.
    """
    year_to_dates = defaultdict(set)
    for d in dates:
        year_to_dates[d.year].add(d)

    full_years = set()
    other_dates = set()

    for year, dates in six.iteritems(year_to_dates):
        year_len = 365 + int(calendar.isleap(year))
        if len(dates) >= year_len:
            full_years.add(year)
        else:
            other_dates.update(dates)

    return sorted(full_years), sorted(other_dates)


def extract_full_months(dates):
    """Find if there are any years where every day of the year is in *dates*.

    Returns ``full_months, other_dates``: *months* is a list of tuples
    of ``(year, month)`` for months that *dates* contains all the days of,
    and *other_dates* is everything else.
    """
    month_to_dates = defaultdict(set)
    for d in dates:
        month_to_dates[(d.year, d.month)].add(d)

    full_months = set()
    other_dates = set()

    for (year, month), dates in six.iteritems(month_to_dates):
        month_len = calendar.monthrange(year, month)[1]
        if len(dates) >= month_len:
            full_months.add((year, month))
        else:
            other_dates.update(dates)

    return sorted(full_months), sorted(other_dates)


def extract_full_tens(dates):
    """
    Find if there are any series of dates containing a full range of ten
    dates in *dates*, e.g. 2016-01-0* 2016-01-1* 2016-01-2* 2016-01-3*.

    :type dates: list
    :param dates: list of dates to find ten-day periods in.
    :return full_tens: a set of dates that can be globbed into ten-day periods
    :return other_dates: remaining dates that cannot be globbed
    """
    ten_to_dates = defaultdict(set)

    for d in dates:
        ten_to_dates[(d.year, d.month, which_ten(d))].add(d)

    full_tens = set()
    other_dates = set()

    for (year, month, ten), dates in six.iteritems(ten_to_dates):
        mr = calendar.monthrange(year, month)[1]
        if ten < 3 and len(dates) == TEN_LENGTH[ten]:
            full_tens.add((year, month, ten))
        elif ten == 3 and len(dates) == mr - 30 + 1:  # Handle 30+ days
            full_tens.add((year, month, ten))
        elif month == 2 and ten == 2 and len(dates) == mr - 20 + 1:  # Handle February
            full_tens.add((year, month, ten))
        else:
            other_dates.update(dates)

    return sorted(full_tens), sorted(other_dates)


def which_ten(d):
    """
    Determine which ten-day period a date falls in.

    :type d: datetime.date
    :param d: day of date, e.g. 2016-01-02 in format %Y-%m-%d is 2
    :return: an integer that is the prefix of the relevant ten-day period.
    """
    if not calendar.monthrange(d.year, d.month)[1] >= d.day > 0:
        raise RuntimeError('Out of range date')

    if d.day < 10:
        return 0
    elif d.day < 20:
        return 1
    elif d.day < 30:
        return 2
    else:
        return 3
